walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -14,10 +14,220 @@
+      
+    String endOfDayTitle = "END_OF_DAY";
+    String startOfDayTitle = "START_OF_DAY";
+    Event endOfDay = new Event(endOfDayTitle, TimeRange.fromStartDuration(TimeRange.END_OF_DAY + 1, 0), allAttendees);
+    Event startOfDay = new Event(startOfDayTitle, TimeRange.fromStartDuration(TimeRange.START_OF_DAY, 0), allAttendees);
+
+    LinkedList<Event> sortedEvents = new LinkedList<Event>(events);
+
+    sortedEvents.add(endOfDay);
+    sortedEvents.add(startOfDay);
+      
+    sortedEvents.sort(Event.EVENT_COMPARATOR);
+
+    LinkedList<Event> applicableEventsOnly = removeEventsWithNoAttendees(mandatoryAttendees, sortedEvents);
+    LinkedList<Event> mandatoryNonOverlapping = sortedEventsIntoNonOverlappingSortedEvents(mandatoryAttendees, applicableEventsOnly);
+    ArrayList<TimeRange> available = availableTimesFromSortedNonOverlappingEvents(mandatoryNonOverlapping, request);
+
+    if (optionalAttendees.size() == 0) {
+      return available;
+    }
+
+    LinkedList<Event> applicableEventsOnlyOfOptional = removeEventsWithNoAttendees(optionalAttendees, sortedEvents);
+    LinkedList<Event> optionalNonOverlapping = sortedEventsIntoNonOverlappingSortedEvents(optionalAttendees, applicableEventsOnlyOfOptional);
+    ArrayList<TimeRange> availableWithOptional = availableTimesFromSortedNonOverlappingEvents(optionalNonOverlapping, request);
+    
+    if (available.size() == 0) {
+      return availableWithOptional;
+    }
+    
+    ArrayList<TimeRange> finalTimeRanges = mergeTimeRangesThatIntersect(availableWithOptional, available);
+    return finalTimeRanges;
+  }
+   
+  /*
+   * Function to merge time ranges that intersect between two lists of time ranges
+   * 
+   * Runtime: O(n*logm) where 'n' is the length of optional and 'm' is the length of mandatory
+   * 
+   * @param optional    A list of time ranges in which for every one we'll search for an intersection
+   * @param mandatory   A list of time ranges to search within
+   * 
+   * return   A list of time ranges where all ranges are within both input lists
+   * 
+  */
+  private ArrayList<TimeRange> mergeTimeRangesThatIntersect(ArrayList<TimeRange> optional, ArrayList<TimeRange> mandatory) {
+    ArrayList<TimeRange> finalTimeRanges = new ArrayList<>();
+
+    ArrayList<Integer> startValues = new ArrayList<>();
+    for (TimeRange mandatoryTimeRange: mandatory) {
+      startValues.add(mandatoryTimeRange.start());
+    }
+
+    for (TimeRange optionalTimeRange: optional) {
+      int start = optionalTimeRange.start();
+      int end = optionalTimeRange.end();
+
+      // This only works because java's binary search returns (-(insertion_point) - 1) 
+      // if not found, meaning that the negative is 2 indexes beyond what the closest 
+      // lesser than element exists
+      int index = Collections.binarySearch(startValues, start);
+      index = index >= 0 ? index : (index + 2) * -1;
```

Comment:
Hmm, I think this should still be `(index + 1) * -1`

If you think about this case:
`[1, 2, 3, 4]`, `binarySearch` for `5` will return `-4`
`(-4 + 2) * -1 = 2`, which is not the current insertion index.

I think it would make more sense have `index` be the correct insertion point, and change L119 to check for `index != 0`

Reply:
For that case it return -5 actually. The way I see it the insertion index is where the key would go, so in this case it would go in index 4. Simply doing the inverse of -insertion -1 gives you the closest biggest element. Also why would L119 be index != 0?

Reply:
Oh, you're correct, it would be `-5`. But with `(index + 1) * -1`, that's what you want actually
`(-5 +1) * -1 = 4`, which is the 0-based index insertion point for `5`.

If we were inserting `0`, you'd get -1
`(-1 + 1) * -1 = 0`, which is the 0-based index insertion point for `0`.

And with your algorithm, you reject 0 as an insertion point, because your 0th element is always START_OF_DAY, so inserting something before that doesn't make sense.

Reply:
And none of this is relevant anymore of course :) but just adding thoughts for completion.

Reply:
ohhhh yeah I see that